src/routes/zome_call.rs

```diff
@@ -55,53 +77,127 @@ pub async fn zome_call(
     State(state): State<AppState>,
     Query(query): Query<PayloadQuery>,
 ) -> HcHttpGatewayResult<()> {
-    check_payload_size(
-        query.payload.as_ref(),
-        state.configuration.payload_limit_bytes,
-    )?;
-
-    let _decoded_payload = if let Some(payload) = query.payload {
-        let decoded = BASE64_URL_SAFE.decode(payload)?;
-        let json = serde_json::from_slice::<serde_json::Value>(&decoded)?;
-        Some(json)
-    } else {
-        None
-    };
-    Ok(())
-}
-
-fn check_payload_size(
-    payload: Option<&String>,
-    payload_limit_bytes: u32,
-) -> HcHttpGatewayResult<()> {
-    if let Some(encoded_payload) = payload {
-        let estimated_decoded_size = calculate_base64_decoded_size(encoded_payload);
-
-        if estimated_decoded_size > payload_limit_bytes {
+    // Check payload byte length does not exceed configured maximum.
+    if let Some(payload) = query.payload {
+        if payload.as_bytes().len() > state.configuration.payload_limit_bytes as usize {
             return Err(HcHttpGatewayError::PayloadSizeLimitError {
-                size: estimated_decoded_size,
-                limit: payload_limit_bytes,
+                size: payload.as_bytes().len() as u32,
+                limit: state.configuration.payload_limit_bytes,
             });
         }
+
+        // Decode payload to JSON.
+        let _zome_call_payload = base64_json_to_hsb(&payload)?;
     }
 
     Ok(())
 }
 
-/// Calculate the approximate decoded size without actually decoding
-/// Base64 encoding: every 4 chars in base64 represent 3 bytes of original data
-/// Need to account for padding characters too ('='), which don't represent data
-fn calculate_base64_decoded_size(encoded_payload: &str) -> u32 {
```

Comment:
Actually I think now that we need to check the byte length of the base64 encoded payload, not the length of the decoded payload. The spec says this limit is supposed to restrict the overall request size.